modules/dem_uart/test/test_dem_uart.py

```diff
@@ -0,0 +1,222 @@
+"""
+    test_dem_uart
+    
+    Cocotb-based unit test for UART-Device Emulation Module (DEM UART)
+"""
+import random
+import cocotb
+
+from cocotb.clock import Clock
+from cocotb.triggers import Timer, RisingEdge
+
+from osdtestlib.debug_interconnect import RegAccess, DiPacket, NocDiReader, NocDiWriter
+from osdtestlib.asserts import *
+from osdtestlib.exceptions import DebugEventFailedException
+
+# Debug Interconnect address for the DEM module
+MODULE_DI_ADDRESS = 1
+
+# Debug interconnect address for the testing module
+SENDER_DI_ADDRESS = 0
+
+
+@cocotb.coroutine
+def _init_dut(dut):
+    
+    #Setup clock
+    cocotb.fork(Clock(dut.clk, 1000).start())
+
+    # Reset DUT
+    dut._log.debug("Resetting DUT");
+    dut.rst <= 1
+    
+    dut.id <= MODULE_DI_ADDRESS
+    
+    for _ in range(2):
+        yield RisingEdge(dut.clk)
+    dut.rst <= 0    
+
+@cocotb.test()
+def test_dem_uart_baseregisters(dut):
+    """
+    Verify that the base configuration registers have the correct values.
+    """
+    
+    access = RegAccess(dut)
+    
+    yield _init_dut(dut)
+    
+    yield access.test_base_registers(MODULE_DI_ADDRESS, SENDER_DI_ADDRESS,
+                                     mod_vendor=1, mod_type=5, mod_version=0,
+                                     can_stall=True)
+
+@cocotb.coroutine
+def _activate_module(dut):
+    """
+    Set the MOD_CS_ACTIVE bit in the Control and Status register to 1 to
+    enable transmission of debug event packets.
+    """
+
+    access = RegAccess(dut)
+
+    yield access.write_register(dest=MODULE_DI_ADDRESS,
+                                src=SENDER_DI_ADDRESS,
+                                word_width=16,
+                                regaddr=DiPacket.BASE_REG.MOD_CS.value,
+                                value=1)
+
+@cocotb.test()
+def test_dem_uart_activation(dut):
+    """
+    Check if DEM_UART is handling the activation bit correctly
+    """
+
+    access = RegAccess(dut)
+
+    yield _init_dut(dut)
+
+    dut._log.info("Check contents of MOD_CS")
+    yield access.assert_reg_value(MODULE_DI_ADDRESS, SENDER_DI_ADDRESS,
+                                  DiPacket.BASE_REG.MOD_CS.value, 16, 0)
+
+    yield _activate_module(dut)
+
+    dut._log.info("Check contents of MOD_CS")
+    yield access.assert_reg_value(MODULE_DI_ADDRESS, SENDER_DI_ADDRESS,
+                                  DiPacket.BASE_REG.MOD_CS.value, 16, 1)
+
+# Mirroring expected transfer through module
+_fifo = []
+
+@cocotb.test()
+def test_dem_to_dii(dut):
+    """
+    Verify that no chars are lost when the DII is busy
+    """
+
+    yield _init_dut(dut)
+    yield _activate_module(dut)
+
+    write_thread = cocotb.fork(dem_to_dii_tx(dut, num_transfers=1000, random_data=True))
+    read_thread  = cocotb.fork(dem_to_dii_rx(dut, num_transfers=1000))
```

Comment:
Random testing is a very nice thing to explore corner cases, such as backpressure conditions. But to make sure that every test run covers those issues, you should add explicit tests for those cases to your test. E.g. add a test in which the NoC is never ready, and test what happens.  Even though this scenario is (most likely, depending on the random number generator) already covered, making corner cases explicit significantly helps during development.